src/qt/textctrl.cpp

```diff
@@ -14,19 +14,364 @@
+};
+
+class wxQtSingleLineEdit : public wxQtEdit
+{
+public:
+    wxQtSingleLineEdit(QLineEdit *edit) :
+        edit(edit)
+    {
+    }
+
+    virtual bool IsModified() const wxOVERRIDE
+    {
+        return edit->isModified();
+    }
+
+    virtual int GetNumberOfLines() const wxOVERRIDE
+    {
+        return 1;
+    }
+
+    virtual wxString DoGetValue() const wxOVERRIDE
+    {
+        return wxQtConvertString( edit->text() );
+    }
+
+    virtual long GetInsertionPoint() const wxOVERRIDE
+    {
+        long selectionStart = edit->selectionStart();
+
+        if(selectionStart >= 0)
+            return selectionStart;
+
+        return edit->cursorPosition();
+    }
+
+    virtual QWidget *GetHandle() const wxOVERRIDE
+    {
+        return edit;
+    }
+
+    virtual int GetLineLength(long WXUNUSED(lineNo)) const wxOVERRIDE
+    {
+        return DoGetValue().length();
+    }
+
+    virtual wxString GetLineText(long lineNo) const wxOVERRIDE
+    {
+        return lineNo == 0 ? DoGetValue() : wxString();
+    }
+
+    virtual void WriteText( const wxString &text ) wxOVERRIDE
+    {
+        edit->insert(wxQtConvertString( text ));
+    }
+
+    virtual void MarkDirty() wxOVERRIDE
+    {
+        return edit->setModified( true );
+    }
+
+    virtual void DiscardEdits() wxOVERRIDE
+    {
+        return edit->setModified( false );
+    }
+
+    virtual void blockSignals(bool block) wxOVERRIDE
+    {
+        edit->blockSignals(block);
+    }
+
+    virtual void SetValue( const wxString &value ) wxOVERRIDE
+    {
+        edit->setText(wxQtConvertString( value ));
+    }
+
+    virtual void SetSelection( long from, long to ) wxOVERRIDE
+    {
+        edit->setSelection(from, to - from);
+    }
+
+    virtual bool GetSelection( long *from, long *to ) const wxOVERRIDE
+    {
+        *from = edit->selectionStart();
+        if ( *from < 0 ) return false;
+
+        *to = *from + edit->selectedText().length();
+        return true;
+    }
+
+    virtual void SetInsertionPoint(long pos) wxOVERRIDE
+    {
+        // check if pos indicates end of text:
+        if (pos == -1)
+            edit->end(false);
+        else
+            edit->setCursorPosition(pos);
+    }
+
+    virtual long XYToPosition(long x, long y) const wxOVERRIDE
+    {
+        if(y == 0 && x >= 0)
+        {
+            if(static_cast<size_t>(x) <= DoGetValue().length()) return x;
+        }
+
+        return -1;
+    }
+
+    virtual bool PositionToXY(long pos, long *x, long *y) const wxOVERRIDE
+    {
+        const wxString &value = DoGetValue();
+
+        if(static_cast<size_t>(pos) > value.length()) return false;
+
+        *y = 0;
+        *x = pos;
+        return true;
+    }
+
+    virtual QScrollArea *ScrollBarsContainer() const wxOVERRIDE
+    {
+        return NULL;
+    }
+
+private:
+    QLineEdit *edit;
+};
+
+wxQtEdit *createwxQtEdit(wxWindow *parent, wxTextCtrl *textCtrl, long style)
+{
+    bool multiline = (style & wxTE_MULTILINE) != 0;
```

Comment:
Could just use `textCtrl->IsMultiLine()` instead

Reply:
Unfortunately, this would not work here. The Create function sets up the m_qtEdit before calling QtCreateControl, which sets up base. The QtCreateControl call depends on m_qtEdit being set up first (temporal coupling), and will segfault if it is not. It could probably be fixed, but it is a bigger amount of work. I'd rather have the tests passing first before tackling something like this.